### PR TITLE
Add support for DataFrames and Numpy Arrays for log_plot()

### PR DESCRIPTION
**Pull request by mnrozhkov**
This PR is a part of this issue: https://github.com/iterative/dvclive/issues/750 

Changes: 
- add support for Add support for DataFrames and Numpy Arrays in `log_plot()` 
- implemented a new function for convert datapoints to List[Dict] `dvclive/utils.py/convert_datapoints_to_list_of_dicts` 
- added new Python dependencies: `pandas` and `numpy`
- added a new argument to the `log_plot()`: `columns: Optional[List[str]] = None,` - optional list of column names for data in Numpy Array format

JN to test new feature: [log_plots.ipynb](https://drive.google.com/file/d/1tDoigZrpo_YzPiS-fXsrDHSJuSmgM-uT/view?usp=sharing)

### TODO
- [x] ❗ I have followed the [Contributing to DVCLive](https://github.com/iterative/dvclive/blob/main/CONTRIBUTING.rst)
  guide.
- [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR: https://github.com/iterative/dvc.org/pull/5055

### CODE REVIEW

File: src/dvclive/live.py [hunk 1]
```diff
@@ -2,7 +2,9 @@
 import json
 import logging
 import math
+import numpy as np
 import os
+import pandas as pd
```
Comment (dberenbaum) [inline src/dvclive/live.py:7]: Can we make these optional dependencies and add them here?

https://github.com/iterative/dvclive/blob/84069208144be711ef0feccc38064652d732eac7/pyproject.toml#L46

You can still handle type checking here without always importing those deps: https://stackoverflow.com/questions/61384752/how-to-type-hint-with-an-optional-import

File: src/dvclive/live.py [hunk 2]
```diff
@@ -391,14 +394,20 @@ def log_image(self, name: str, val):
...
         template: Optional[str] = None,
         title: Optional[str] = None,
         x_label: Optional[str] = None,
         y_label: Optional[str] = None,
+        columns: Optional[List[str]] = None,
     ):
+        # Convert the given datapoints to List[Dict]
+        datapoints = convert_datapoints_to_list_of_dicts(
+            datapoints=datapoints, columns=columns
+        )
```
Comment (dberenbaum) [inline src/dvclive/live.py:409]: Can we check here if it's a list first and skip in that case? That way, you can still plot a list without needing pandas and numpy installed.

Comment (dberenbaum) [reply src/dvclive/live.py:409]: Or you can leave this and check it within that function itself. For example:

https://github.com/iterative/dvclive/blob/84069208144be711ef0feccc38064652d732eac7/src/dvclive/live.py#L372-L389

File: src/dvclive/utils.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
 import csv
 import json
+import numpy as np
 import os
+import pandas as pd
```
Comment (dberenbaum) [inline src/dvclive/utils.py:5]: Let's import these only as needed inside the function itself. See the example above from `log_image()`.

File: src/dvclive/live.py [hunk 4]
```diff
@@ -391,14 +394,20 @@ def log_image(self, name: str, val):
...
         name: str,
-        datapoints: List[Dict],
+        datapoints: pd.DataFrame | np.ndarray | List[Dict],
         x: str,
         y: str,
         template: Optional[str] = None,
         title: Optional[str] = None,
         x_label: Optional[str] = None,
         y_label: Optional[str] = None,
+        columns: Optional[List[str]] = None,
```
Comment (dberenbaum) [inline src/dvclive/live.py:404]: I'm not sure we need this. It's only useful for the specific scenario of labeling numpy array columns AFAICT, and there are already several ways to handle that which are at least as straightforward IMO:

1. Convert to a pd dataframe yourself
2. Use [structured arrays](https://numpy.org/doc/stable/user/basics.rec.html)

**Review by dberenbaum — CHANGES_REQUESTED** [review]
Thanks @mnrozhkov! Very clean and simple. I have just two overall suggestions:

1. Try to require numpy and pandas as narrowly as possible
2. Drop `columns` since I think users can achieve the same without it

File: src/dvclive/live.py (hunk 2 above)
Comment (mnrozhkov) [reply src/dvclive/live.py:409]: Ok, it makes sense!  Should we raise an error if none of the supported formats `List[Dict] |pd.DataFrame | np.ndarray`  provided?

Comment (dberenbaum) [reply src/dvclive/live.py:409]: I guess an error will already get raised below:

https://github.com/iterative/dvclive/blob/19b71b4c05c0e24f747ee4df4e331f960a4dbd97/src/dvclive/live.py#L402

However, I don't think it's a bad idea to check the types here and raise if it's not one of those. Not a blocker to me however you want to handle it.

**Comment by mnrozhkov** [comment]
> Thanks @mnrozhkov! Very clean and simple. I have just two overall suggestions:
> 
> 1. Try to require numpy and pandas as narrowly as possible
> 2. Drop `columns` since I think users can achieve the same without it

Thanks @dberenbaum! I resolved all comments. Pls, let me know if there are other places to improve 🙌

File: src/dvclive/utils.py [hunk 5]
```diff
@@ -194,3 +199,39 @@ def read_history(live, metric):
...
+    Raises:
+        ValueError: If the `datapoints` argument is not of type pd.DataFrame, np.ndarray, or List[Dict].
+    """
+    if isinstance(datapoints, list):
+        return datapoints
+
+    import pandas as pd
+
+    if isinstance(datapoints, pd.DataFrame):
+        return datapoints.to_dict(orient="records")
```
Comment (dberenbaum) [inline src/dvclive/utils.py:225]: I think you could use this:

https://github.com/iterative/dvclive/blob/19b71b4c05c0e24f747ee4df4e331f960a4dbd97/src/dvclive/utils.py#L153-L157

That way you can check instance types without importing the packages. Hopefully that means we would only import pandas when passed a pandas dataframe.

**Comment by dberenbaum** [comment]
Thanks @mnrozhkov! Left one last comment and then I think it's ready to merge. Do you plan to also do the docs PR?

**Comment by mnrozhkov** [comment]
> Thanks @mnrozhkov! Left one last comment and then I think it's ready to merge. Do you plan to also do the docs PR?

Yes, I plan to update the docs as well.

File: src/dvclive/utils.py (hunk 5 above)
Comment (skshetry) [reply src/dvclive/utils.py:225]: I'll suggest importing at the top and suppressing on import error.

```python
# top level

try:
   import pandas as pd
except ImportError:
   pd = None
```

And guard/check for None.
```python
if pd and isinstance(datapoints, pd.DataFrame):
    pass
```

Comment (dberenbaum) [reply src/dvclive/utils.py:225]: @mnrozhkov I'll defer to @skshetry here as the engineer.

@skshetry would you advise we drop `is_instance_without_import()` (in a separate PR)?

Comment (mnrozhkov) [reply src/dvclive/utils.py]: @skshetry, thanks for the suggestions! I dropped `is_instance_without_import()` and made it work with this imports: 

```python
if TYPE_CHECKING:
    import numpy as np
    import pandas as pd
else:
    try:
        import pandas as pd
    except ImportError:
        pd = None

    try:
        import numpy as np
    except ImportError:
        np = None
```

It doesn't look nice, but it's the only solution works to meet 2 goals 
- use proper type annotations for `datapoints` 
- meet `mypy` checks

Comment (skshetry) [reply src/dvclive/utils.py]: It's okay to silence mypy here here.

File: src/dvclive/utils.py [hunk 6]
```diff
@@ -1,12 +1,31 @@
...
+        pd = None
+
+    try:
+        import numpy as np
+    except ImportError:
+        np = None
+
+from .error import (
+    InvalidDataTypeError,
+)
```
Comment (skshetry) [inline src/dvclive/utils.py:28]: Please move this before `if TYPE_CHECKING` block. Also, adding a comma at the end hangs this, so you can remove it and keep it in a single line.
```suggestion
from .error import InvalidDataTypeError
```

File: tests/test_utils.py [hunk 7]
```diff
@@ -15,3 +18,31 @@
 )
 def test_standardize_metric_name(framework, logged, standardized):
     assert standardize_metric_name(logged, framework) == standardized
+
+
+class TestConvertDatapointsToListOfDicts:
```
Comment (skshetry) [inline tests/test_utils.py:23]: Please see if you can use `parametrize` here for these four tests.
Also there will be no need to put them into a class after that, as they will drop to just 2 tests: one for valid outputs and other for invalid outputs.

**Review by dberenbaum — APPROVED** [review]
Thanks @mnrozhkov! Great improvement here.

**Pull request merged**